src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1382,26 +1371,56 @@ JvmtiEnvBase::check_top_frame(JavaThread* current_thread, JavaThread* java_threa
 
 jvmtiError
 JvmtiEnvBase::force_early_return(JavaThread* java_thread, jvalue value, TosState tos) {
-  JavaThread* current_thread = JavaThread::current();
-  HandleMark   hm(current_thread);
-  uint32_t debug_bits = 0;
-
   // retrieve or create the state
   JvmtiThreadState* state = JvmtiThreadState::state_for(java_thread);
   if (state == NULL) {
     return JVMTI_ERROR_THREAD_NOT_ALIVE;
   }
 
-  // Check if java_thread is fully suspended
-  if (!java_thread->is_thread_fully_suspended(true /* wait for suspend completion */, &debug_bits)) {
-    return JVMTI_ERROR_THREAD_NOT_SUSPENDED;
+  // Eagerly reallocate scalar replaced objects.
+  JavaThread* current_thread = JavaThread::current();
+  EscapeBarrier eb(true, current_thread, java_thread);
+  if (eb.barrier_active()) {
+    if (java_thread->frames_to_pop_failed_realloc() > 0) {
+      // VM is in the process of popping the top frame because it has scalar replaced objects
+      // which could not be reallocated on the heap.
+      // Return JVMTI_ERROR_OUT_OF_MEMORY to avoid interfering with the VM.
+      return JVMTI_ERROR_OUT_OF_MEMORY;
+    }
+    if (!eb.deoptimize_objects(0)) {
```

Comment:
@reinrich did I mess something up when merging this in?

Reply:
Stack frames are counted beginning at 0. The top frame has depth 0. So object deoptimization happens in the top frame.

Still the used method is not optimal because it assumes that objects of frames within the given depth are accessed and their escape state is changed. But potentially caller methods optimized on the escape state therefore it searches for caller frames passing ArgEscape objects and deoptimizes these too. With ForceEarlyReturn no objects are accessed but it is so uncommon that I did not bother optimizing this. Should I?

Reply:
@robehn you haven't messed up. Hope I havn't either. I've tested

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/hotspot/jtreg:hotspot_serviceability     197   197     0     0   
   jtreg:test/jdk:jdk_svc                             1176  1176     0     0   
   jtreg:test/jdk:jdk_jdi                              174   174     0     0   
   jtreg:test/hotspot/jtreg:vmTestbase_nsk_jdi        1141  1141     0     0   
   jtreg:test/hotspot/jtreg:vmTestbase_nsk_jvmti       648   648     0     0   
   jtreg:test/hotspot/jtreg:vmTestbase_nsk_jdwp        113   113     0     0   
==============================
TEST SUCCESS
```
jdk_jdi now includes jdk/com/sun/jdi/EATests.java which tests PopFrame/ForceEarlyReturn with object reallocation with and without reallocation failures.

Reply:
Ah. I see now the loop uses <= instead of <. So my reasoning was right but off by 1. Passing in 0 really means deopt 1 frame. Which means everything is fine and working the way I expect it to.

Reply:
Great